I3uckwheat/cards-of-carousal
# Bug/250 duplicate keys in player lists

#### 0. Make sure your issue is linked:

Close #250 

#### 1. Purpose:

1. Refactor PlayerList to not use names as keys
2. Refactor PlayerKicker and OptionList to not use player names as keys
3. Prevent players from using the name of a player currently in the game / staging going forward. 


#### 2. Implementation:

1.
- Swap the players name with finding that players ID and using that as the key
2. 
- Convert OptionList to take an array of objects with a `text` and a `value` rather than just an array of strings. The `text` is the text that should be rendered on the accordion fold, the `value` is what will be passed to the `onItemClick` callback when that fold is clicked. 
- Update PlayerKicker to match this new API.

3. 

- Put the HostContext event handling switch into a thunk to make state accessible from within it.
- Based off of the state conditionally dispatch the `PLAYER_CONNECTED` action as before, or (if the player attempting to join has used the same name as a player already in the game or in staging) dispatch a new `SEND_NAME_TAKEN_MESSAGE` action.
- Add this new action to the HostReducerMiddleware and have it update the `gameState` of the offending player to the new `connection-refused-name-taken` state.
- Update `PlayerScreenController` to inform the offending player that they have been refused access to the lobby when this state is received, and explain why they were refused using the `PlayerErrorScreen` with some snark baked in for flavor. 


#### 3. Possible Issues:




#### 4. How to Test:


0. Fetch my branch and switch to it
1. Start the application backend then frontend
2. Open at least three browser windows
3. Select host mode on one and join on the others
4. Attempt to use the same name for both players
5. Verify that you are not able to join the lobby using the name of a player already in staging
6. Verify that you *are* able to join the lobby using a name not yet in use
7. Start the game
8. Open another browser window and select join mode again
9. Attempt to join using the name of one of your players already in the game
10. Verify that you are not able to join the lobby using the name of a player already in the game
11. Verify that you *are* able to join the lobby using a name not yet in use
12. Using dev tools verify that the PlayerList and PlayerKicker are using playerId as the key now rather than name 


#### 5. Cleanup
